test/Kestrel.FunctionalTests/Http2/ShutdownTests.cs

```diff
@@ -0,0 +1,125 @@
+                    listenOptions.Protocols = HttpProtocols.Http2;
+                    listenOptions.UseHttps(_x509Certificate2);
+                });
+            }))
+            {
+                var requestTask = Client.GetStringAsync($"https://localhost:{server.Port}/");
+                Assert.False(requestTask.IsCompleted);
+
+                await requestStarted.Task.DefaultTimeout();
+
+                var stopTask = server.StopAsync();
+
+                // Unblock the request
+                requestUnblocked.SetResult(null);
+
+                Assert.Equal("hello world HTTP/2", await requestTask);
+                Assert.Equal(stopTask, await Task.WhenAny(stopTask, Task.Delay(10000)));
+            }
+
+            Assert.Contains(TestApplicationErrorLogger.Messages, m => m.Message.Contains("Request finished in"));
+            Assert.Contains(TestApplicationErrorLogger.Messages, m => m.Message.Contains("is closing."));
+            Assert.Contains(TestApplicationErrorLogger.Messages, m => m.Message.Contains("is closed. The last processed stream ID was 1."));
+        }
+
+        [ConditionalFact]
+        public async Task GracefulTurnsAbortiveIfRequestsDoNotFinish()
+        {
+            var requestStarted = new TaskCompletionSource<object>();
+            var requestUnblocked = new TaskCompletionSource<object>();
+            // Abortive shutdown leaves one request hanging
+            using (var server = new TestServer(TransportSelector.GetWebHostBuilder(new DiagnosticMemoryPoolFactory(allowLateReturn: true).Create), async context =>
+            {
+                requestStarted.SetResult(null);
+                await requestUnblocked.Task.DefaultTimeout();
+                await context.Response.WriteAsync("hello world " + context.Request.Protocol);
+            }, new TestServiceContext(LoggerFactory),
+            kestrelOptions =>
+            {
+                kestrelOptions.Listen(IPAddress.Loopback, 0, listenOptions =>
+                {
+                    listenOptions.Protocols = HttpProtocols.Http2;
+                    listenOptions.UseHttps(_x509Certificate2);
+                });
+            },
+            _ => { }))
+            {
+                var requestTask = Client.GetStringAsync($"https://localhost:{server.Port}/");
+                Assert.False(requestTask.IsCompleted);
+                await requestStarted.Task.DefaultTimeout();
+
+                var stopTask = server.StopAsync();
+
+                // Keep the request blocked
+                Assert.Equal(stopTask, await Task.WhenAny(stopTask, Task.Delay(10000)));
+            }
+
+            Assert.Contains(TestApplicationErrorLogger.Messages, m => m.Message.Contains("is closing."));
+            Assert.Contains(TestApplicationErrorLogger.Messages, m => m.Message.Contains("is closed. The last processed stream ID was 1."));
+            Assert.DoesNotContain(TestApplicationErrorLogger.Messages, m => m.Message.Contains("Request finished in"));
```

Comment:
Does "Some connections failed to close gracefully during server shutdown." get logged in this test?